### PR TITLE
feat: add ci-asp workflow

### PR DESCRIPTION
**Pull request by udaltsovra**
The workflow builds the application on ASP.NET and deploys it to Azure Cloud

### CODE REVIEW

File: .github/workflows/ci-asp-app.yaml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    # More information for connect your workflows to authenticate with Azure.
+    # https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure
+
+    permissions:
+      id-token: write
+      contents: read
+
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
```
Comment (MikhailKorol-saritasa) [inline .github/workflows/ci-asp-app.yaml:66]:
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/ci-asp-app.yaml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+      - name: Add msbuild to PATH
+        uses: microsoft/setup-msbuild@v1.1
```
Comment (MikhailKorol-saritasa) [inline .github/workflows/ci-asp-app.yaml:69]:
```suggestion
        uses: microsoft/setup-msbuild@v1.3
```

File: .github/workflows/ci-asp-app.yaml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      
+      - name: Setup NuGet
+        uses: NuGet/setup-nuget@v1.0.5
```
Comment (MikhailKorol-saritasa) [inline .github/workflows/ci-asp-app.yaml:72]:
```suggestion
        uses: NuGet/setup-nuget@v1.2
```

File: .github/workflows/ci-asp-app.yaml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        with:
+          name: ASP-app
+          path: ${{ inputs.BUILD_PATH }}/${{ inputs.BUILD_FOLDER }}/${{ inputs.APP_NAME }}-${{ inputs.environment }}
+      
+  # Generated by Azure Depoloyment Center
+  # Docs for the Azure Web Apps Deploy action: https://github.com/Azure/webapps-deploy
+  # More GitHub Actions for Azure: https://github.com/Azure/actions
+  
+  # Deploying an application to azure cloud
+  deploy:
```
Comment (MikhailKorol-saritasa) [inline .github/workflows/ci-asp-app.yaml:107]: Mb it's better to use step with condition instead separated job?

**Review by MikhailKorol-saritasa — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-asp-app.yaml (hunk 1 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: Done

File: .github/workflows/ci-asp-app.yaml (hunk 2 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: Done

File: .github/workflows/ci-asp-app.yaml (hunk 3 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: Done

File: .github/workflows/ci-asp-app.yaml (hunk 4 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:107]: It seems to me more convenient when the assembly and deployment are in different jobs, it is convenient to look at the logs in the gh

File: .github/workflows/ci-asp-app.yaml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+# - deploy to Azure App Service
+# 
+# Original Author: Rostislav Udaltsov
+# Created: 05/11/2023
+# ----------------------------------------------------------------------------------------------------------
+
+on:
+  workflow_call:
+    inputs:
+      environment:
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:16]: - use lowercase for all vars.
- add description to each and every value.

File: .github/workflows/ci-asp-app.yaml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      AZURE_TENANT_ID:
+        required: true
+      publish-profile:
+        required: true
+        
+jobs:
+  
+  # Installing dependencies and building the application
+  build:
+    runs-on: windows-latest
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:54]: that should be parametersized too from the workflow.

File: .github/workflows/ci-asp-app.yaml [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        run: .\rattus-windows-amd64.exe > ${{ inputs.BUILD_PATH }}/Web.Debug.config
+        env:
+          AZURE_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
+          AZURE_CLIENT_SECRET: ${{ secrets.AZURE_CLIENT_SECRET }}
+          AZURE_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
+          AZURE_VAULT: ${{ inputs.AZURE_VAULT }}
+          TEMPLATE_PATH: ${{ inputs.TEMPLATE_PATH }}
+     
+      - name: Build app for release
+        run: msbuild ${{ inputs.BUILD_PATH }}/${{ inputs.CSPROJ_FILE }} /p:Configuration=Debug /p:Platform=AnyCPU /t:WebPublish /p:WebPublishMethod=FileSystem /p:DeleteExistingFiles=True /p:publishUrl=${{ inputs.BUILD_FOLDER }}/${{ inputs.APP_NAME }}-${{ inputs.environment }}
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:94]: use multiline

File: .github/workflows/ci-asp-app.yaml [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+          name: ASP-app
+          path: ${{ inputs.BUILD_PATH }}/${{ inputs.BUILD_FOLDER }}/${{ inputs.APP_NAME }}-${{ inputs.environment }}
+      
+  # Generated by Azure Depoloyment Center
+  # Docs for the Azure Web Apps Deploy action: https://github.com/Azure/webapps-deploy
+  # More GitHub Actions for Azure: https://github.com/Azure/actions
+  
+  # Deploying an application to azure cloud
+  deploy:
+    runs-on: windows-latest
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:108]: parameterize this.

File: .github/workflows/ci-asp-app.yaml [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    needs: build
+    environment:
+      name: 'Develop'
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:111]: parameterize this.

File: .github/workflows/ci-asp-app.yaml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      url: ${{ steps.deploy-to-webapp.outputs.webapp-url }}
+    
+    steps:
+      - name: Download artifact from build job
+        uses: actions/download-artifact@v2
+        with:
+          name: ASP-app
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:118]: parameterize this

File: .github/workflows/ci-asp-app.yaml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+      # A publish profile is a file that contains information and settings for deploy applications and services in Azure Cloud
+      # publish-profile is a file which is unique for each app serivce. It can be downloaded from app service's overview page on Azure portal:
+      # https://docs.microsoft.com/en-us/azure/app-service/deploy-github-actions?tabs=applevel#generate-deployment-credentials
+      
+      - name: Deploy to Azure Web App
+        id: deploy-to-webapp
+        uses: azure/webapps-deploy@v2
+        with:
+          app-name: '${{ inputs.APP_NAME }}-${{ inputs.environment }}'
+          slot-name: 'production' # deployment slot by default for new application
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:130]: parameterize this

File: .github/workflows/ci-asp-app.yaml [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+          # Get publish-profile from https://keys.saritasa.cloud/cred/detail/16857/
+          publish-profile: ${{ secrets.publish-profile }}
+          package: .
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:133]: add slack notification on success/failure.

File: .github/workflows/ci-asp-app.yaml [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+      # publish-profile is a file which is unique for each app serivce. It can be downloaded from app service's overview page on Azure portal:
+      # https://docs.microsoft.com/en-us/azure/app-service/deploy-github-actions?tabs=applevel#generate-deployment-credentials
+      
+      - name: Deploy to Azure Web App
+        id: deploy-to-webapp
+        uses: azure/webapps-deploy@v2
+        with:
+          app-name: '${{ inputs.APP_NAME }}-${{ inputs.environment }}'
+          slot-name: 'production' # deployment slot by default for new application
+          # Get publish-profile from https://keys.saritasa.cloud/cred/detail/16857/
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:131]: that should not be here - instead - it should be in the ci-dev.yml in the concrete repository calling this workflow.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
in addition pls create docs in saritasa-devops-docs and explain
- entire workflow how to manage the app
- setup/regeneate secrets
- what secrets are required, where to get them, how to regenerate them
- what values (attrs passed in the workflow) are required, what they mean, how to get the values etc
- record the video using vimeo, https://keys.saritasa.cloud/cred/detail/13328/

Add your tutorial here
https://github.com/saritasa-nest/saritasa-devops-docs/tree/main/_solutions

name it "azure web services"

File: .github/workflows/ci-asp-app.yaml (hunk 6 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:54]: done

File: .github/workflows/ci-asp-app.yaml (hunk 8 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:108]: done

File: .github/workflows/ci-asp-app.yaml (hunk 10 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:118]: done

File: .github/workflows/ci-asp-app.yaml (hunk 9 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: done

File: .github/workflows/ci-asp-app.yaml (hunk 11 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:133]: done

File: .github/workflows/ci-asp-app.yaml (hunk 13 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:134]: done

File: .github/workflows/ci-asp-app.yaml (hunk 4 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: Combined jobs build and deploy

File: .github/workflows/ci-asp-app.yaml (hunk 5 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:16]: done

File: .github/workflows/ci-asp-app.yaml (hunk 12 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml:137]: image

done

File: .github/workflows/ci-asp-app.yaml (hunk 7 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: done

**Comment by udaltsovra** [comment]
> in addition pls create docs in saritasa-devops-docs and explain
> 
> * entire workflow how to manage the app
> * setup/regeneate secrets
> * what secrets are required, where to get them, how to regenerate them
> * what values (attrs passed in the workflow) are required, what they mean, how to get the values etc
> * record the video using vimeo, https://keys.saritasa.cloud/cred/detail/13328/
> 
> Add your tutorial here https://github.com/saritasa-nest/saritasa-devops-docs/tree/main/_solutions
> 
> name it "azure web services"

Okay, I'll do it.

**Review by MikhailKorol-saritasa — APPROVED** [review]

**Review by populov — APPROVED** [review]

File: .github/workflows/ci-asp-app.yaml [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        required: true
+        type: string
+        description: 'Deployment slot in Azure Cloud'
+      runner_image:
+        required: true
+        type: string
+        description: 'The image of the runner OS on which workflow is running'
+    # For get slack channel id use this instruction
+    # https://www.wikihow.com/Find-a-Channel-ID-on-Slack-on-PC-or-Mac
+      slack_channel_id:
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:68]: use https://github.com/marketplace/actions/slack-send as random slack channel ID is impossible to test and hard to find the name by ID.

get the slack webhook URL
https://api.slack.com/apps/A01LM626QTZ/incoming-webhooks

Comment (dmitry-mightydevops) [reply .github/workflows/ci-asp-app.yaml:68]: add a link above in the docs - so we can quickly openit.

File: .github/workflows/ci-asp-app.yaml [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+      
+      - name: Upload artifact for deployment task
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ inputs.app_name }}
+          path: ${{ inputs.build_path }}/${{ inputs.build_folder }}/${{ inputs.app_name }}-${{ inputs.environment }}
+      
+  # Generated by Azure Depoloyment Center
+  # Docs for the Azure Web Apps Deploy action: https://github.com/Azure/webapps-deploy
+  # More GitHub Actions for Azure: https://github.com/Azure/actions
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci-asp-app.yaml:158]: wrong padding on the left

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-asp-app.yaml (hunk 14 above)
Comment (dmitry-mightydevops) [reply .github/workflows/ci-asp-app.yaml:68]: actually disregard it - as we're going to keep the webhook in the secret - it won't be readable at the code level anyway

**Review by dmitry-mightydevops — APPROVED** [review]

File: .github/workflows/ci-asp-app.yaml (hunk 15 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: done

File: .github/workflows/ci-asp-app.yaml (hunk 14 above)
Comment (udaltsovra) [reply .github/workflows/ci-asp-app.yaml]: image

done

**Pull request merged**